src/backend/distributed/utils/multi_partitioning_utils.c

```diff
@@ -37,6 +46,248 @@
+	}
+
+	return taskList;
+}
+
+
+/*
+ * ConstraintNameListForShard returns a list of constraint names for a shard that needs to
+ *  be renamed. All of the following should be satisfied:
+ *	- The constraint name has a shardId suffix.
+ *	- Constraint type is different than foreign key or a unique.
+ *
+ * Partitions inherit constraints from their parents. If a partitioned table shard has a
+ * shardId suffix in a constraint name, then that constraint is inherited in the child
+ * shards as well. As the shardId of the child is  different than that of the parent, it
+ * can cause issues. (See #3970)
+ *
+ * Note that we skip all foreign key and unique constraints, as the single caller of this
+ * function needs the list of constraints that should be renamed, and foreign keys and
+ * unique constraints should always have shardId suffixes in their names to remain unique
+ * in the schema.
+ *
+ * PostgreSQL requires constraint names to be unique among the constraints of a
+ * single relation or domain. This is enforced by a unique index on
+ * conrelid + contypid + conname on pg_constraint. However there are two more limitations:
+ *	- UNIQUE constraints create UNIQUE indexes in the schema, and indexes should have
+ *	  unique names. That's the reason why it is not allowed to have 2 UNIQUE constraints
+ *	  on different tables with the same name.
+ *	- Similarly FOREIGN KEY constraints also create indexes that need unique names on the
+ *	  schema.
+ */
+static List *
+ConstraintNameListForShard(Oid relationId, int32 shardId)
+{
+	List *constraintNameList = NIL;
+
+	int scanKeyCount = 1;
+	ScanKeyData scanKey[1];
+
+	Relation pgConstraint = table_open(ConstraintRelationId, AccessShareLock);
+
+	ScanKeyInit(&scanKey[0], Anum_pg_constraint_conrelid,
+				BTEqualStrategyNumber, F_OIDEQ, relationId);
+
+	bool useIndex = true;
+	SysScanDesc scanDescriptor = systable_beginscan(pgConstraint,
+													ConstraintRelidTypidNameIndexId,
+													useIndex, NULL, scanKeyCount,
+													scanKey);
+
+	HeapTuple heapTuple = systable_getnext(scanDescriptor);
+	while (HeapTupleIsValid(heapTuple))
+	{
+		Form_pg_constraint constraintForm = (Form_pg_constraint) GETSTRUCT(heapTuple);
+		char *constraintName = NameStr(constraintForm->conname);
+
+		/* skip foreign key and unique constraints */
+		if (constraintForm->contype != CONSTRAINT_FOREIGN &&
+			constraintForm->contype != CONSTRAINT_UNIQUE &&
+			NameHasShardIdSuffix(constraintName, shardId))
+		{
+			constraintNameList = lappend(constraintNameList, pstrdup(constraintName));
+		}
+
+		heapTuple = systable_getnext(scanDescriptor);
+	}
+
+	systable_endscan(scanDescriptor);
+	table_close(pgConstraint, NoLock);
+
+	return constraintNameList;
+}
+
+
+/*
+ * RemoveShardIdFromConstraintNameCommand creates the command that will remove the
+ * shardId suffix from a constraint name.
+ */
+static char *
+RemoveShardIdFromConstraintNameCommand(Oid relationId, char *constraintName,
+									   int32 shardId)
```

Comment:
would this correctly handle cases where the constraint name is very long and gets truncated with a hash?

e.g. 

`ck_012345678901234567890123456789012345678901234567890123456789` becomes `ck_01234567890123456789012345678901234567890123_8478db72_102008`

Reply:
I added test cases for this scenario in both regression and upgrade tests. They both seem fine.

`worker_fix_pre_citus10_partitioned_table_constraint_names()` now accepts a new parameter for the original constraint name so that we can pass that information to the worker nodes.